vespa-athenz/src/main/java/com/yahoo/vespa/athenz/identityprovider/AthenzIdentityProviderImpl.java

```diff
@@ -129,96 +114,66 @@ private boolean isExpired(AthenzCredentials credentials) {
     }
 
     private static Instant getExpirationTime(AthenzCredentials credentials) {
-        return credentials.getCreatedAt().plus(EXPIRES_AFTER).minus(EXPIRATION_MARGIN);
+        return credentials.getCertificate().getNotAfter().toInstant();
     }
 
-    private class UpdateCredentialsTask implements RunnableWithTag {
-        @Override
-        public void run() {
-            try {
-                AthenzCredentials newCredentials = isExpired(credentials)
-                        ? athenzCredentialsService.registerInstance()
-                        : athenzCredentialsService.updateCredentials(credentials);
-                credentials = newCredentials;
-                scheduler.schedule(new UpdateCredentialsTask(), UPDATE_PERIOD);
-            } catch (Throwable t) {
-                log.log(LogLevel.WARNING, "Failed to update credentials: " + t.getMessage(), t);
-                lastThrowable.set(t);
-                Duration timeToExpiration = Duration.between(clock.instant(), getExpirationTime(credentials));
-                // NOTE: Update period might be after timeToExpiration, still we do not want to DDoS Athenz.
-                Duration updatePeriod =
-                        timeToExpiration.compareTo(UPDATE_PERIOD) > 0 ? UPDATE_PERIOD : REDUCED_UPDATE_PERIOD;
-                scheduler.schedule(new UpdateCredentialsTask(), updatePeriod);
-            }
-        }
-
-        @Override
-        public String tag() {
-            return UPDATE_CREDENTIALS_TAG;
+    void refreshCertificate() {
+        try {
+            AthenzCredentials newCredentials = isExpired(credentials)
+                    ? athenzCredentialsService.registerInstance()
+                    : athenzCredentialsService.updateCredentials(credentials);
+            credentials = newCredentials;
+        } catch (Throwable t) {
+            log.log(LogLevel.WARNING, "Failed to update credentials: " + t.getMessage(), t);
         }
     }
 
-    private class CertificateExpiryMetricUpdater implements RunnableWithTag {
-        private final Metric metric;
-
-        private CertificateExpiryMetricUpdater(Metric metric) {
-            this.metric = metric;
-        }
-
-        @Override
-        public void run() {
+    void reportMetrics() {
+        try {
             Instant expirationTime = getExpirationTime(credentials);
             Duration remainingLifetime = Duration.between(clock.instant(), expirationTime);
             metric.set(CERTIFICATE_EXPIRY_METRIC_NAME, remainingLifetime.getSeconds(), null);
-            scheduler.schedule(this, CERTIFICATE_EXPIRY_METRIC_UPDATE_PERIOD);
+        } catch (Throwable t) {
+            log.log(LogLevel.WARNING, "Failed to update metrics: " + t.getMessage(), t);
         }
+    }
 
+    private class UpdateCredentialsTask implements Runnable {
```

Comment:
Consider replacing the `UpdateCredetialsTask` and `CertificateExpiryMetricUpdater` classes with a simple method reference when invoking `scheduleAtFixedRate()`.